lib/block_view_validator.go

```diff
@@ -786,21 +793,62 @@ func DBPutValidatorWithTxn(
 		return nil
 	}
 
-	// Set ValidatorEntry in PrefixValidatorByPKID.
+	// Look up the existing ValidatorEntry from the db
+	dbEntry, err := DBGetValidatorByPKIDWithTxn(txn, snap, validatorEntry.ValidatorPKID)
+	if err != nil {
+		return errors.Wrapf(err, "DBUpdateValidatorWithTxn: ")
+	}
+	dbEntryBytes := EncodeToBytes(blockHeight, dbEntry)
+
+	entryToWriteBytes := EncodeToBytes(blockHeight, validatorEntry)
+	// If the entry we're about to write is the exact same as what's already in the db then
+	// don't write it.
+	//
+	// In 99%+ of cases, the entries will be identical so we save a lot from
+	// this optimization, and it significantly speeds up block processing ot have it. When they
+	// differ, typically it's only because of LastActiveAtEpochNumber. For this reason, we have
+	// a secondary optimization to only update the PrefixValidatorByStatusAndStakeAmount index
+	// when absolutely necessary.
+	if bytes.Equal(dbEntryBytes, entryToWriteBytes) {
+		return nil
+	}
+
+	// Set ValidatorEntry in PrefixValidatorByPKID. This should gracefully overwrite an existing entry
+	// if one exists.
 	key := DBKeyForValidatorByPKID(validatorEntry)
 	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry), eventManager); err != nil {
 		return errors.Wrapf(
 			err, "DBPutValidatorWithTxn: problem storing ValidatorEntry in index PrefixValidatorByPKID",
 		)
 	}
 
-	// Set ValidatorEntry key in PrefixValidatorByStatusAndStakeAmount. The value should be nil.
-	// We parse the ValidatorPKID from the key for this index.
-	key = DBKeyForValidatorByStatusAndStakeAmount(validatorEntry)
-	if err := DBSetWithTxn(txn, snap, key, nil, eventManager); err != nil {
-		return errors.Wrapf(
-			err, "DBPutValidatorWithTxn: problem storing ValidatorEntry in index PrefixValidatorByStatusAndStakeAmount",
-		)
+	// If the entry we're about to write has the exact same stake amount and the exact same status,
+	// then there is no need to update PrefixValidatorByStatusAndStakeAmount. This saves us a lot in terms
+	// of block processing time due to the aforementioned badger bug.
+	if dbEntry == nil || validatorEntry.TotalStakeAmountNanos.Cmp(dbEntry.TotalStakeAmountNanos) != 0 ||
+		validatorEntry.Status() != dbEntry.Status() {
```

Comment:
Same thing here, we need to check if the remaining validatorEntry fields have change: `Domains`, `DisableDelegatedStake`, `DelegatedStakeCommissionBasisPoints`.

Otherwise, the validator will never be able to update their domain or other properties once the entry has been created

Reply:
This only applies to `PrefixValidatorByStatusAndStakeAmount` which is just Prefix + Status + total stake amount nanos + validator PKID, so we technically don't need to check the other fields.